reconcile/utils/terraform_client.py

```diff
@@ -531,34 +531,62 @@ def cleanup(self):
+        # replacement and thus it's not included here, see:
+        #   https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_RenameInstance.html
+        supported_values: dict[str, str] = {
+            "engine_version": "EngineVersion",
+            "storage_type": "StorageType",
+            "allocated_storage": "AllocatedStorage",
+            "instance_class": "DBInstanceClass",
+            "multi_az": "MultiAZ",
+            "iops": "Iops",
+        }
         before = resource_change["before"]
         after = resource_change["after"]
-        changed_terraform_args = [
+        changed_resource_arguments = [
             key for key, value in before.items() if value != after.get(key)
         ]
-        if (
-            len(changed_terraform_args) == 1
-            and "engine_version" in changed_terraform_args
-            and self._aws_api is not None
-        ):
+        if changed_resource_arguments and self._aws_api is not None:
             region_name = get_region_from_availability_zone(before["availability_zone"])
             response = self._aws_api.describe_rds_db_instance(
                 account_name, resource_name, region_name=region_name
             )
-            pending_modified_values = response["DBInstances"][0][
-                "PendingModifiedValues"
-            ]
-            if (
-                "EngineVersion" in pending_modified_values
-                and pending_modified_values["EngineVersion"]
-                == resource_change["after"]["engine_version"]
-            ):
-                return True
-        return False
+            try:
+                pending_modified_values = response["DBInstances"][0].get(
+                    "PendingModifiedValues"
+                )
+            except (KeyError, IndexError):
+                raise RuntimeError(f"Expected RDS database {resource_name} to exist")
+            if pending_modified_values is None:
+                # The PendingModifiedValues attribute might not be included as
+                # part of the RDS database instance object. However, we expect
+                # it to be included here, especially since, at this point,
+                # Terraform also claims a pending change for the underlying
+                # resource. Thus if we can't agree on AWS API vs Terraform
+                # state, it's better to raise to signify a potential problem.
+                raise RuntimeError(
+                    f"Expected RDS database {resource_name} to have pending changes"
+                )
+            pending_values: list[str] = []
+            for argument in changed_resource_arguments:
+                value = supported_values[argument]
+                if (
+                    value in pending_modified_values
+                    and cast(dict[str, str], pending_modified_values)[value]
+                    == after[argument]
+                ):
+                    pending_values.append(value)
+        return pending_values
```

Comment:
This changes the behavior of this method from returning `True` or `False`, indicating that all the changes in `changed_resource_arguments` have already been applied and are accounted for in `PendingModifiedValues` to only now checking if at least 1 is?

To make this more concrete, imagine that I've done the following:

1. I change `EngineVersion` and run `terraform-resources`
2. The AWS API accepted the change and now `EngineVersion` is in `PendingModifiedValues`
3. I now change `AllocatedStorage` and run `terraform-resources`

The expected behavior would be that the `apply` would run because the diff in `changed_resource_arguments` doesn't exactly match the changes queued in `PendingModifiedValues`?

Tell me if I've misunderstood this and I'll re-read the code again rather than you needing to spell it out for me :)

Reply:
@steveteahan, I have updated the pending changes detection logic to allow for incoming intermediate changes.

Assuming that **apply_immediately** is still disabled, any additional new change to the resource on Terraform's side that does not have a corresponding entry in the set we obtain from **PendingModifiedValues**, will enable apply to take place - this should result in the changes to be scheduled on the AWS' side, and then for us to detect a matching set of pending changes on both sides for a subsequent reconciliation run.

Let me know what you think about this approach.